### ISSUE TITLE
Intermittent hang with docker exec near container exit

### ISSUE THREAD

**Issue by leth** [docker/docker#27626]
We intermittently see `docker exec ...` hanging when we use it near container process exit.
We see this both [with the python SDK](https://github.com/docker/docker-py/issues/1269) and the cli tool. 

**Steps to reproduce the issue:**

Run this test harness for a while:

``` python
import time
import itertools
from subprocess import check_output, CalledProcessError


def docker(*args):
    return check_output(('docker', ) + args)

for i in itertools.count():
    print(i, time.time())
    container = docker(
        'run', '--net=host', '-d', 'busybox:latest', 'sh', '-c', 'sleep 0.55'
    ).strip()
    try:
        docker('exec', container, 'sh', '-c', 'killall sleep || true')
    except CalledProcessError as e:
        pass
    docker('wait', container)
    docker('rm', container)
```

**Describe the results you received:**
Eventually hangs

**Describe the results you expected:**
Should not hang

**Output of `docker version`:**

```
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:29:41 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:29:41 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 7
 Running: 0
 Paused: 0
 Stopped: 7
Images: 195
Server Version: 1.12.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 481
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-34-generic
Operating System: Linux Mint 18
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.679 GiB
Name: kohlenstoff
ID: 2TCK:YUZU:NWIV:QETB:XR42:UC3N:SNR6:MP4E:UAZ3:VT6C:TMKA:UWIS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running on my laptop!

**Comment by mschwartz** [comment]
I have had frequent issues shutting down containers.  The daemon goes away and I have to kill all the processes or factory reset docker to use docker again.

```
docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.7.1
 Git commit:   6f9534c
 Built:        Thu Sep  8 10:31:18 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64
```

**Comment by cpuguy83** [comment]
ping @mlaventure

**Comment by mlaventure** [comment]
@mschwartz Your issue sounds different, you may want to open a new issue with more details

**Comment by mlaventure** [comment]
@leth could you send SIGUSR1 to the daemon and provide the daemon log afterwards?

Thanks

**Comment by mschwartz** [comment]
https://github.com/docker/docker/issues/27635

Next time it crashed/hangs, I'll do the SIGUSR1 signal and add to the above issue.

The good news is that the containers ran over the weekend and didn't crash.  That's in spite of them crashing daily the week before.

I haven't updated any of the software.

**Comment by mschwartz** [comment]
For the life of me, I can't tell which of the dozen+ processes with "docker" in the name is the "daemon."

```
ps -ax | grep docker
 9145 ??         0:00.35 /Applications/Docker.app/Contents/MacOS/com.docker.osx.hyperkit.linux -watchdog fd:0
 9147 ??         0:00.20 /Applications/Docker.app/Contents/MacOS/com.docker.osx.hyperkit.linux -watchdog fd:0
91413 ??         0:00.58 com.docker.db --url fd:3 --git /Users/mschwartz/Library/Containers/com.docker.docker/Data/database
91414 ??         0:00.04 com.docker.osx.hyperkit.linux
91417 ??         0:00.01 /Applications/Docker.app/Contents/MacOS/com.docker.osx.hyperkit.linux
91811 ??         0:37.84 com.docker.osxfs --address fd:3 --connect /Users/mschwartz/Library/Containers/com.docker.docker/Data/@connect --control fd:4 --volume-control fd:5 --database /Users/mschwartz/Library/Containers/com.docker.docker/Data/s40
91812 ??         0:00.02 com.docker.slirp --db /Users/mschwartz/Library/Containers/com.docker.docker/Data/s40 --ethernet fd:3 --port fd:4 --vsock-path /Users/mschwartz/Library/Containers/com.docker.docker/Data/@connect --max-connections 900
91813 ??         0:00.07 com.docker.driver.amd64-linux -db /Users/mschwartz/Library/Containers/com.docker.docker/Data/s40 -osxfs-volume /Users/mschwartz/Library/Containers/com.docker.docker/Data/s30 -slirp /Users/mschwartz/Library/Containers/com.docker.docker/Data/s50 -vmnet /var/tmp/com.docker.vmnetd.socket -port /Users/mschwartz/Library/Containers/com.docker.docker/Data/s51 -vsock /Users/mschwartz/Library/Containers/com.docker.docker/Data -docker /Users/mschwartz/Library/Containers/com.docker.docker/Data/s60 -addr fd:3 -debug
91814 ??         0:00.02 /Applications/Docker.app/Contents/MacOS/com.docker.driver.amd64-linux -db /Users/mschwartz/Library/Containers/com.docker.docker/Data/s40 -osxfs-volume /Users/mschwartz/Library/Containers/com.docker.docker/Data/s30 -slirp /Users/mschwartz/Library/Containers/com.docker.docker/Data/s50 -vmnet /var/tmp/com.docker.vmnetd.socket -port /Users/mschwartz/Library/Containers/com.docker.docker/Data/s51 -vsock /Users/mschwartz/Library/Containers/com.docker.docker/Data -docker /Users/mschwartz/Library/Containers/com.docker.docker/Data/s60 -addr fd:3 -debug
91816 ??         1:15.94 /Applications/Docker.app/Contents/MacOS/com.docker.hyperkit -A -m 2G -c 4 -u -s 0:0,hostbridge -s 31,lpc -s 2:0,virtio-vpnkit,uuid=bd118f46-d40d-4bb4-8618-e22dc210be4b,path=/Users/mschwartz/Library/Containers/com.docker.docker/Data/s50,macfile=/Users/mschwartz/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/mac.0 -s 3,virtio-blk,file:///Users/mschwartz/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/Docker.qcow2,format=qcow -s 4,virtio-9p,path=/Users/mschwartz/Library/Containers/com.docker.docker/Data/s40,tag=db -s 5,virtio-rnd -s 6,virtio-9p,path=/Users/mschwartz/Library/Containers/com.docker.docker/Data/s51,tag=port -s 7,virtio-sock,guest_cid=3,path=/Users/mschwartz/Library/Containers/com.docker.docker/Data,guest_forwards=2376;1525 -l com1,autopty=/Users/mschwartz/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/tty,log=/Users/mschwartz/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/console-ring -f kexec,/Applications/Docker.app/Contents/Resources/moby/vmlinuz64,/Applications/Docker.app/Contents/Resources/moby/initrd.img,earlyprintk=serial console=ttyS0 com.docker.driver="com.docker.driver.amd64-linux", com.docker.database="com.docker.driver.amd64-linux" ntp=gateway mobyplatform=mac -F /Users/mschwartz/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/hypervisor.pid
95202 ??         0:00.01 /Applications/Docker.app/Contents/MacOS/com.docker.frontend {"action":"vmstateevent","args":{"vmstate":"running"}}
95203 ??         0:00.01 /Applications/Docker.app/Contents/MacOS/com.docker.frontend {"action":"vmstateevent","args":{"vmstate":"running"}}
95205 ttys002    0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn docker
```

**Comment by mschwartz** [comment]
![image](https://cloud.githubusercontent.com/assets/329637/19667441/c49918b6-9a05-11e6-9a98-249d0c23ad54.png)

**Comment by mlaventure** [comment]
@mschwartz there  should  be a `/var/run/docker.pid` file with the daemon `PID` :)

**Comment by mschwartz** [comment]
There is no /var/run/docker.pid file.  There is a /var/run/docker.sock file (I assume a unix domain socket).

It happened again.

```
web_1    | Tue, 25 Oct 2016 00:50:03 GMT index 200 GET /bootstrap/dist/css/bootstrap.min.css.map
mongo_1  | 2016-10-25T00:50:33.426+0000 I NETWORK  [conn132] end connection 172.18.0.3:52532 (3 connections now open)
mongo_1  | 2016-10-25T00:50:33.899+0000 I NETWORK  [conn130] end connection 172.18.0.3:52518 (2 connections now open)
mongo_1  | 2016-10-25T02:51:52.010+0000 W FTDC     [ftdc] Uncaught exception in 'FileRenameFailed: Software caused connection abort' in full-time diagnostic data capture subsystem. Shutting down the full-time diagnostic data capture subsystem.
mongo_1  | 2016-10-25T02:51:52.583+0000 E STORAGE  [thread1] WiredTiger (107) [1477363912:583798][1:0x7f16e6d45700], log-server: /data/db/journal: directory-list: opendir: Transport endpoint is not connected
mongo_1  | 2016-10-25T02:51:52.583+0000 E STORAGE  [thread1] WiredTiger (107) [1477363912:583906][1:0x7f16e6d45700], log-server: log pre-alloc server error: Transport endpoint is not connected
mongo_1  | 2016-10-25T02:51:52.583+0000 E STORAGE  [thread1] WiredTiger (107) [1477363912:583914][1:0x7f16e6d45700], log-server: log server error: Transport endpoint is not connected
mongo_1  | 2016-10-25T02:51:52.606+0000 E STORAGE  [thread2] WiredTiger (107) [1477363912:606898][1:0x7f16e5542700], file:WiredTiger.wt, WT_SESSION.checkpoint: /data/db/WiredTiger.turtle: file-exist: stat: Transport endpoint is not connected
mongo_1  | 2016-10-25T02:51:52.606+0000 E STORAGE  [thread2] WiredTiger (107) [1477363912:606956][1:0x7f16e5542700], checkpoint-server: checkpoint server error: Transport endpoint is not connected
mongo_1  | 2016-10-25T02:51:52.606+0000 E STORAGE  [thread2] WiredTiger (-31804) [1477363912:606962][1:0x7f16e5542700], checkpoint-server: the process must exit and restart: WT_PANIC: WiredTiger library panic
mongo_1  | 2016-10-25T02:51:52.606+0000 I -        [thread2] Fatal Assertion 28558
mongo_1  | 2016-10-25T02:51:52.606+0000 I -        [thread2]
mongo_1  |
mongo_1  | ***aborting after fassert() failure
mongo_1  |
mongo_1  |
mongo_1  | 2016-10-25T02:51:52.621+0000 F -        [thread2] Got signal: 6 (Aborted).
mongo_1  |
mongo_1  |  0x13287d2 0x1327909 0x1328112 0x7f16e89038d0 0x7f16e857e067 0x7f16e857f448 0x12ae512 0x10a74a3 0x96f386 0x96f45b 0x96f61f 0x1a10deb 0x7f16e88fc0a4 0x7f16e863162d
mongo_1  | ----- BEGIN BACKTRACE -----
mongo_1  | {"backtrace":[{"b":"400000","o":"F287D2","s":"_ZN5mongo15printStackTraceERSo"},{"b":"400000","o":"F27909"},{"b":"400000","o":"F28112"},{"b":"7F16E88F4000","o":"F8D0"},{"b":"7F16E8549000","o":"35067","s":"gsignal"},{"b":"7F16E8549000","o":"36448","s":"abort"},{"b":"400000","o":"EAE512","s":"_ZN5mongo13fassertFailedEi"},{"b":"400000","o":"CA74A3"},{"b":"400000","o":"56F386","s":"__wt_eventv"},{"b":"400000","o":"56F45B","s":"__wt_err"},{"b":"400000","o":"56F61F","s":"__wt_panic"},{"b":"400000","o":"1610DEB"},{"b":"7F16E88F4000","o":"80A4"},{"b":"7F16E8549000","o":"E862D","s":"clone"}],"processInfo":{ "mongodbVersion" : "3.2.10", "gitVersion" : "79d9b3ab5ce20f51c272b4411202710a082d0317", "compiledModules" : [], "uname" : { "sysname" : "Linux", "release" : "4.4.20-moby", "version" : "#1 SMP Thu Sep 15 12:10:20 UTC 2016", "machine" : "x86_64" }, "somap" : [ { "elfType" : 2, "b" : "400000", "buildId" : "03AADFF60B81BAF2675CA42BB00380AB31F376AF" }, { "b" : "7FFE481A0000", "path" : "linux-vdso.so.1", "elfType" : 3, "buildId" : "62A6D37117F69E399C5567F03BF8B6BC5D1F2AA3" }, { "b" : "7F16E9830000", "path" : "/usr/lib/x86_64-linux-gnu/libssl.so.1.0.0", "elfType" : 3, "buildId" : "0711D67B299ACA3EA6B0B57C0DA4EC83B71787EE" }, { "b" : "7F16E9434000", "path" : "/usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.0", "elfType" : 3, "buildId" : "2E613A282B22F2B0978ACCEC1B71411F5E20CC89" }, { "b" : "7F16E922C000", "path" : "/lib/x86_64-linux-gnu/librt.so.1", "elfType" : 3, "buildId" : "18650C365FBA12BA12DB680B988B7754C0F99364" }, { "b" : "7F16E9028000", "path" : "/lib/x86_64-linux-gnu/libdl.so.2", "elfType" : 3, "buildId" : "1D841026FF2C5AEA25C21921FAC9C978AB465825" }, { "b" : "7F16E8D27000", "path" : "/lib/x86_64-linux-gnu/libm.so.6", "elfType" : 3, "buildId" : "BBA5CF06072F15BEA36FC8B0A2C6A32BBF08C235" }, { "b" : "7F16E8B11000", "path" : "/lib/x86_64-linux-gnu/libgcc_s.so.1", "elfType" : 3, "buildId" : "D5FB04F64B3DAEA6D6B68B5E8B9D4D2BC1A6E1FC" }, { "b" : "7F16E88F4000", "path" : "/lib/x86_64-linux-gnu/libpthread.so.0", "elfType" : 3, "buildId" : "B9DD2674891C5395C5724321158DC2EB8C7DA64B" }, { "b" : "7F16E8549000", "path" : "/lib/x86_64-linux-gnu/libc.so.6", "elfType" : 3, "buildId" : "AD0CE0A6B6E201C9A8195B2D55043AFCD4F5DB00" }, { "b" : "7F16E9A91000", "path" : "/lib64/ld-linux-x86-64.so.2", "elfType" : 3, "buildId" : "44DA6E33EBAB324C1D0128F9465425AD355CC978" } ] }}
mongo_1  |  mongod(_ZN5mongo15printStackTraceERSo+0x32) [0x13287d2]
mongo_1  |  mongod(+0xF27909) [0x1327909]
mongo_1  |  mongod(+0xF28112) [0x1328112]
mongo_1  |  libpthread.so.0(+0xF8D0) [0x7f16e89038d0]
mongo_1  |  libc.so.6(gsignal+0x37) [0x7f16e857e067]
mongo_1  |  libc.so.6(abort+0x148) [0x7f16e857f448]
mongo_1  |  mongod(_ZN5mongo13fassertFailedEi+0x82) [0x12ae512]
mongo_1  |  mongod(+0xCA74A3) [0x10a74a3]
mongo_1  |  mongod(__wt_eventv+0x3F4) [0x96f386]
mongo_1  |  mongod(__wt_err+0x8D) [0x96f45b]
mongo_1  |  mongod(__wt_panic+0x24) [0x96f61f]
mongo_1  |  mongod(+0x1610DEB) [0x1a10deb]
mongo_1  |  libpthread.so.0(+0x80A4) [0x7f16e88fc0a4]
mongo_1  |  libc.so.6(clone+0x6D) [0x7f16e863162d]
mongo_1  | -----  END BACKTRACE  -----
mbreanodejs_mongo_1 exited with code 14
Exception in thread Thread-4:
Traceback (most recent call last):
  File "threading.py", line 810, in __bootstrap_inner
  File "threading.py", line 763, in run
  File "compose/cli/log_printer.py", line 149, in tail_container_logs
  File "compose/cli/log_printer.py", line 179, in wait_on_exit
  File "compose/container.py", line 239, in wait
  File "site-packages/docker/utils/decorators.py", line 21, in wrapped
  File "site-packages/docker/api/container.py", line 441, in wait
  File "site-packages/docker/client.py", line 174, in _raise_for_status
APIError: 502 Server Error: Bad Gateway ("Bad response from Docker engine")
```

**Comment by mlaventure** [comment]
Oh, this is D4M.

ping @justincormack what's the way to get the stacktrace in that context?

**Comment by mschwartz** [comment]
I really hope this isn't an issue in production on a linux host.

**Comment by leth** [comment]
Guys, we have two likely different issues here, on two different platforms. Can we keep discussion of the other issue on #27635 otherwise everyone is going to get confused.

**Comment by leth** [comment]
IIRC on Mac docker runs in a VM using the built-in virtualization tools

On 25 October 2016 00:22:02 BST, Michael Schwartz <notifications@github.com> wrote:
>![image](https://cloud.githubusercontent.com/assets/329637/19667441/c49918b6-9a05-11e6-9a98-249d0c23ad54.png)
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/docker/docker/issues/27626#issuecomment-255893336

-- 
Sent from my phone. Please excuse my brevity.

**Comment by leth** [comment]
Sorry, I'm out of the office for the next week. The test harness should be able to reproduce the issue in the mean time.

On 24 October 2016 17:17:50 BST, "Kenfe-Mickaël Laventure" <notifications@github.com> wrote:
>@leth could you send SIGUSR1 to the daemon and provide the daemon log
>afterwards?
>
>Thanks
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/docker/docker/issues/27626#issuecomment-255788332

-- 
Sent from my phone. Please excuse my brevity.

**Comment by leth** [comment]
Here's the [log output from systemd, including the goroutine dump from SIGUSR1](http://pastebin.com/CgBcB6Qn).

**Comment by mlaventure** [comment]
@leth I'd need the daemon logs in debug mode unfortunately, in the current stats there's no enough information for me to have and idea of the flow of operations. Do you think you'd be able to reproduce the issue with that config?

But the daemon is indeed blocked waiting on the exec context/streams to be closed.

It may be that the container already died and we didn't correctly received or took care of the exec death/cancellation, but it's hard to tell.

**Comment by leth** [comment]
Hmm, I'm running it with debug logging on and it hasn't hung yet; It usually doesn't take that long to hang.

Don't you just love race conditions!

**Comment by leth** [comment]
Aha; 978 iterations in, here's the last 1000 lines: http://pastebin.com/Jvy7Eh7U let me know if you need more!

**Comment by mlaventure** [comment]
Thanks!

So, the issue is that we seemed to receive the exec `exit` event after the one from the container `init` process. Hence the container is not "known" by `libcontainerd` at that time.

I thought we had fixed this in `containerd` a while back to ensure exit events were sent in the right order. I'll get another look at the code.

**Comment by mlaventure** [comment]
@leth I had a look, but there's something I'd like to check before making any changes.

I don't seem to be able to find the debug logs from `containerd` into the one you've provided. Are you starting `containerd` separately by any chance?

If not, for some reason the `--debug` flag wasn't passed down to it, and I would need the output to confirm exactly where the error is coming from

**Comment by leth** [comment]
I'm not starting anything weirdly afaik; I only set `--log-level=debug` might that be it?

Have you had any luck with the test script on your machine?

**Comment by mlaventure** [comment]
@leth no luck so far. Will give it another spin though

Regarding the debug mode `--debug` on the daemon commandline is needed for it to be propagated to `containerd`

**Comment by leth** [comment]
I'll give it a go with `--debug` tomorrow!

**Comment by leth** [comment]
The timing fairies were happy today, `--debug` + `--log-level=debug` reproduced it quite quickly: http://pastebin.com/1AsSTaJ9

**Comment by mlaventure** [comment]
@leth thanks!

It confirmed, my suspicion regarding the exit event not being received by `containerd` "in order".

I'll come up with something to sync them.

**Comment by leth** [comment]
Nice work, Thanks! Do you know when this will make it into a docker release?

**Comment by mlaventure** [comment]
Should get there for 1.13.

It reminds me, I need to vendor the new `containerd` in here
